sdk/php/tests/Unit/ValueObject/TypeTest.php

```diff
@@ -0,0 +1,272 @@
+        ];
+
+        yield Dagger\Directory::class => [
+            new Type(Dagger\Directory::class, false),
+            $reflectReturnType(fn(): Dagger\Directory => self
+                ::createStub(Dagger\Directory::class)),
+        ];
+
+        yield sprintf('nullable %s', Dagger\Directory::class) => [
+            new Type(Dagger\Directory::class, true),
+            $reflectReturnType(fn(): ?Dagger\Directory => null),
+        ];
+
+        yield Dagger\File::class => [
+            new Type(Dagger\File::class, false),
+            $reflectReturnType(fn(): Dagger\File => self
+                ::createStub(Dagger\File::class)),
+        ];
+
+        yield sprintf('nullable %s', Dagger\File::class) => [
+            new Type(Dagger\File::class, true),
+            $reflectReturnType(fn(): ?Dagger\File => null),
+        ];
+    }
+
+    /** @return Generator<array{ 0: true, 1:class-string}> */
+    public static function provideIdAbleTypes(): Generator
+    {
+        foreach (self::provideIdAbleClasses() as $idAble) {
+            yield $idAble => [true, $idAble];
+        }
+    }
+
+    /** @return Generator<array{ 0: false, 1:string}> */
+    public static function provideNonIdAbleTypes(): Generator
+    {
+        $nonIdAbles = [
+            'bool',
+            'int',
+            'null',
+            'string',
+            'void',
+            DateTimeImmutable::class,
+        ];
+
+        foreach ($nonIdAbles as $nonIdAble) {
+            yield $nonIdAble => [false, $nonIdAble];
+        }
+    }
+
+    /** @return Generator<array{ 0: Dagger\TypeDefKind, 1:string}> */
+    public static function provideTypeDefKinds(): Generator
+    {
+        yield 'bool' => [Dagger\TypeDefKind::BOOLEAN_KIND, 'bool'];
+        yield 'int' => [Dagger\TypeDefKind::INTEGER_KIND, 'int'];
+        yield 'null' => [Dagger\TypeDefKind::VOID_KIND, 'null'];
+        yield 'string' => [Dagger\TypeDefKind::STRING_KIND, 'string'];
+        yield 'void' => [Dagger\TypeDefKind::VOID_KIND, 'void'];
+        yield DateTimeImmutable::class => [
+            Dagger\TypeDefKind::OBJECT_KIND,
+            DateTimeImmutable::class
+        ];
+
+        yield 'abstract scalar' => [
+            Dagger\TypeDefKind::SCALAR_KIND,
+            Dagger\Platform::class,
+        ];
+
+        foreach (self::provideIdAbleClasses() as $idAbleClass) {
+            yield $idAbleClass => [
+                Dagger\TypeDefKind::OBJECT_KIND,
+                $idAbleClass,
+            ];
+        }
+    }
+
+    /** @return Generator<array{ 0: string, 1:class-string}> */
+    public static function provideShortNames(): Generator
+    {
+        foreach (self::provideIdAbleClasses() as $idAble) {
+            yield $idAble => [
+                (new ReflectionClass($idAble))->getShortName(),
+                $idAble,
+            ];
+        }
+    }
+
+    /** @return class-string[] */
+    private static function provideIdAbleClasses(): array
```

Comment:
Will this function need to be updated if new core types are added?

Reply:
Not if new types are added. Only if existing implementation of `Dagger\Client\Idable` is changed to stop implementing `Dagger\Client\Idable`.

It is a Data Provider for the following test:
https://github.com/dagger/dagger/blob/b11ab3ca47a1e8a1ecf95218f6ec5a09b299468d/sdk/php/tests/Unit/ValueObject/TypeTest.php#L63-L68

As such, it does not need to be an exhaustive list of all implementations, it would probably be fine to only use 3-5 cases.

If we want a dynamically updating list, we could use this:
```php
foreach (get_declared_classes() as $className) {
    if (in_array(Idable::class, class_implements($className))) {
        yield => $className;
    }
}
```

However, looping through every declared class may be a bit excessive.

Reply:
> As such, it does not need to be an exhaustive list of all implementations, it would probably be fine to only use 3-5 cases.

Got it, thanks 👍